src/atlas/interpolation/method/sphericalvector/ComplexMatrixMultiply.h

```diff
@@ -0,0 +1,222 @@
+#include "atlas/parallel/omp/omp.h"
+
+// I don't yet fully trust NVHPC. Turning off OpenMP until we can figure out why
+// it's failing CI.
+#ifdef __NVCOMPILER
+#warning turning off OpenMP for nvhpc build
+#undef atlas_omp_parallel_for
+#define atlas_omp_parallel_for for
+#endif
+
+namespace atlas {
+namespace interpolation {
+namespace method {
+namespace detail {
+
+struct TwoVectorTag {};
+struct ThreeVectorTag {};
+constexpr auto twoVector = TwoVectorTag{};
+constexpr auto threeVector = ThreeVectorTag{};
+
+template <typename VectorTag>
+constexpr bool isVectorTag() {
+  return std::is_same_v<VectorTag, TwoVectorTag> ||
+         std::is_same_v<VectorTag, ThreeVectorTag>;
+}
+
+/// @brief   Helper class to perform complex matrix multiplications
+///
+/// @details Performs matrix multiplication between fields of 2-vectors or
+///          3-vectors. Fields must have Rank >= 2. Here, the assumption is
+///          that Dim = 0 is the horizontal dimension, and Dim = (Rank - 1) is
+///          the vector element dimension.
+template <bool InitialiseTarget>
+class ComplexMatrixMultiply {
+ public:
+  ComplexMatrixMultiply() = default;
+
+  /// @brief   Construct object from sparse matrices.
+  ///
+  /// @details complexWeights is a SparseMatrix of weights. realWeights is a
+  ///          SparseMatrix containing the magnitudes of the elements of
+  ///          complexWeights.
+  ComplexMatrixMultiply(const ComplexMatrix& complexWeights,
+                        const RealMatrix& realWeights)
+      : complexWeightsPtr_{&complexWeights}, realWeightsPtr_{&realWeights} {
+    if constexpr (checkMatrices) {
+      ATLAS_ASSERT(complexWeightsPtr_->rows() == realWeightsPtr_->rows());
+      ATLAS_ASSERT(complexWeightsPtr_->cols() == realWeightsPtr_->cols());
+      ATLAS_ASSERT(complexWeightsPtr_->nonZeros() ==
+                   realWeightsPtr_->nonZeros());
+      for (auto rowIndex = Index{0}; rowIndex < complexWeightsPtr_->rows();
+           ++rowIndex) {
+        for (auto [complexRowIter, realRowIter] = rowIters(rowIndex);
+             complexRowIter; ++complexRowIter, ++realRowIter) {
+          ATLAS_ASSERT(realRowIter);
+          ATLAS_ASSERT(realRowIter.row() == complexRowIter.row());
+          ATLAS_ASSERT(realRowIter.col() == complexRowIter.col());
+          // tinyNum ~= 2.3e-13 for double.
+          constexpr auto tinyNum = 1024 * std::numeric_limits<Real>::epsilon();
+          const auto complexMagnitude = std::abs(complexRowIter.value());
+          const auto realValue = realRowIter.value();
+          const auto error = std::abs(complexMagnitude - realValue);
+
+          const auto printError = [&]() {
+            auto strStream = std::ostringstream{};
+            strStream << "Error complex components: ";
+            strStream << std::setprecision(19) << error;
+            return strStream.str();
+          };
+          ATLAS_ASSERT_MSG(error < tinyNum, printError());
```

Comment:
```suggestion
          ATLAS_ASSERT(error < tinyNum, printError());
```
The ATLAS_ASSERT macro has overload on number of arguments. Second argument is the message.